### PR TITLE
Add withdrawals root to block header

### PR DESCRIPTION
**Pull request by jframe**
## PR description
Add the withdrawalsRoot to the block header.

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/BlockHeader.java [hunk 1]
```diff
@@ -306,6 +316,7 @@ public String toString() {
     sb.append("baseFee=").append(baseFee).append(", ");
     sb.append("mixHashOrPrevRandao=").append(mixHashOrPrevRandao).append(", ");
     sb.append("nonce=").append(nonce);
+    sb.append("withdrawalsRoot=").append(withdrawalsRoot).append(", ");
```
Comment (siladu) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/BlockHeader.java:319]: nit: I mentioned this on Fabio's PR so better mention on here too...should we only be including if it's not null?

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/BlockchainReferenceTestCaseSpec.java [hunk 2]
```diff
@@ -176,7 +176,7 @@ public ReferenceTestBlockHeader(
           baseFee != null ? Wei.fromHexString(baseFee) : null, // baseFee
           Hash.fromHexString(mixHash), // mixHash
           Bytes.fromHexStringLenient(nonce).toLong(),
-          // withdrawalsRoot == null ? Hash.EMPTY : Hash.fromHexString(withdrawalsRoot),
+          withdrawalsRoot != null ? Hash.fromHexString(withdrawalsRoot) : null,
```
Comment (siladu) [inline ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/BlockchainReferenceTestCaseSpec.java:179]: nice!

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestEnv.java [hunk 3]
```diff
@@ -62,7 +62,7 @@ public ReferenceTestEnv(
         Hash.EMPTY, // ommersHash
         Address.fromHexString(coinbase),
         Hash.EMPTY, // stateRoot
-        Hash.EMPTY, // transactionsRoot
+        Hash.EMPTY, // transactionsRoot,
```
Comment (siladu) [inline ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestEnv.java:65]: nit: typo?

**Review by siladu — COMMENTED** [review]
I'm probably misreading but I think I see a mix of Hash.EMPTY and null - are we still using Hash.EMPTY to represent a null withdrawals list?

File: plugin-api/src/main/java/org/hyperledger/besu/plugin/data/BlockHeader.java [hunk 4]
```diff
@@ -70,6 +70,15 @@ public interface BlockHeader {
    */
   Hash getTransactionsRoot();
 
+  /**
+   * The Keccak 256-bit hash of theroot node of the trie structure populated with each withdrawal in
```
Comment (siladu) [inline plugin-api/src/main/java/org/hyperledger/besu/plugin/data/BlockHeader.java:74]: nit: typo s/theroot/the root

File: plugin-api/src/main/java/org/hyperledger/besu/plugin/data/BlockHeader.java [hunk 5]
```diff
@@ -70,6 +70,15 @@ public interface BlockHeader {
...
+   * the withdrawals list portion of the block.
+   *
+   * @return The Keccak 256-bit hash of theroot node of the trie structure populated with each
```
Comment (siladu) [inline plugin-api/src/main/java/org/hyperledger/besu/plugin/data/BlockHeader.java:77]: s/theroot the root

**Comment by jframe** [comment]
> I'm probably misreading but I think I see a mix of Hash.EMPTY and null for the withdrawalsRoot - are we still using Hash.EMPTY to represent a null withdrawals list?

Using null to represent a null withdrawals list. But did use Hash.EMPTY in some test cases where it didn't matter. That is bit misleading I'll just change them all to use null.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/BlockHeader.java (hunk 1 above)
Comment (jframe) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/BlockHeader.java:319]: Yeah that is cleaner

**Review by siladu — APPROVED** [review]
LGTM

**Pull request merged**